code/sonalyze/daemon/perform.go

```diff
@@ -0,0 +1,387 @@
+package daemon
+
+import (
+	"bytes"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"log/syslog"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+	"syscall"
+
+	"go-utils/auth"
+	"go-utils/httpsrv"
+	"go-utils/process"
+)
+
+func (dc *DaemonCommand) RunDaemon(_ io.Reader, _, stderr io.Writer) /* never returns */ {
+	if err := statusStart("jobanalyzer/sonalyzed", stderr); err != nil {
+		fmt.Fprintf(stderr, "FATAL ERROR: Failing to open logger: %s", err.Error())
+		os.Exit(1)
+	}
+
+	// Note "daemon" is not a command here
+	http.HandleFunc("/add", httpAddHandler(dc))
+	http.HandleFunc("/jobs", httpGetHandler(dc, "jobs"))
+	http.HandleFunc("/load", httpGetHandler(dc, "load"))
+	http.HandleFunc("/uptime", httpGetHandler(dc, "uptime"))
+	http.HandleFunc("/profile", httpGetHandler(dc, "profile"))
+	http.HandleFunc("/parse", httpGetHandler(dc, "parse"))
+	http.HandleFunc("/metadata", httpGetHandler(dc, "metadata"))
+	// These request names are compatible with the older `infiltrate` and with the upload infra
+	// already running on the clusters.
+	http.HandleFunc("/sonar-freecsv", httpPostHandler(dc, "sample", "text/csv"))
+	http.HandleFunc("/sysinfo", httpPostHandler(dc, "sysinfo", "application/json"))
+
+	var programFailed bool
+	s := httpsrv.New(dc.Verbose, int(dc.port), func(err error) {
+		programFailed = true
+	})
+	go s.Start()
+
+	// Wait here until we're stopped by SIGHUP (manual) or SIGTERM (from OS during shutdown).
+	//
+	// TODO: IMPROVEME: For SIGHUP, we should not exit but should instead reread the password file
+	// and the cluster aliases file.
+	process.WaitForSignal(syscall.SIGHUP, syscall.SIGTERM)
+	s.Stop()
+
+	if programFailed {
+		os.Exit(1)
+	}
+	os.Exit(0)
+}
+
+// HTTP handlers
+//
+// Documented behavior: the server will close the request body, we don't need to do it.
+//
+// I can find no documentation about needing to consume the body in case of an early (error)
+// return, nor anything obvious in the net/http source code to indicate this, nor has google
+// turned up anything.  So request handler code assumes it's not necessary.
+
+func httpGetHandler(
+	dc *DaemonCommand,
+	command string,
+) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		_, _, clusterName, ok := requestPreamble(dc, w, r, "GET", dc.getAuthenticator, authRealm, "")
+		if !ok {
+			return
+		}
+
+		verb := command
+		arguments := []string{
+			"--data-path",
+			path.Join(dc.jobanalyzerDir, "data", clusterName),
+		}
+		for name, vs := range r.URL.Query() {
+			if name == "cluster" {
+				continue
+			}
+
+			if !argOk(command, name) {
+				w.WriteHeader(400)
+				fmt.Fprintf(w, "Bad parameter %s", name)
+				if dc.Verbose {
+					statusWarningf("Bad parameter %s", name)
+				}
+				return
+			}
+
+			// Repeats are OK, sonalyze allows them in a number of cases.  Booleans carry the magic
```

Comment:
Instead of "sonalyze allows", write "the commands allow".  Maybe look for other references to sonalyze in the daemon code.